### DISCUSSION TITLE
Receiving HID Reports and using Keyboard.h simultaneously

### DISCUSSION THREAD

**Discussion by 2xxn** [earlephilhower/arduino-pico#2317] [General]
I'm kind of new to this topic but what I'm trying to achieve is being able to receive WebHID reports sent like `device.sendReport(ID, PACKET);` in the firmware, I spent around 2 hours looking for a solution and I'm stuck. WebHID already recognizes the device and allows me to connect to it on the site but I can't seem to send reports, failing with `DOMException: Failed to write the report.`

Is it even possible to do with arduino-pico? 

My current code looks something like this
```cpp
#include <Arduino.h>
#include <Keyboard.h>

void setup() {
    Serial.begin(9600);
    Keyboard.begin();
    Keyboard.setAutoReport(false);
}

void loop() {
    // Receive any reports sent via WebHID and handle accordingly
}
```

**Comment by earlephilhower** [comment]
Sorry, I have even less experience than you with WebHID.  However, the Pico itself should have **nothing** to do with it.  It's just a combination CDC/HID device in this case.  If WebHID is trying to send reports to the PC's OS, then the issue is between the WebHID implementation and the OS.  If it's trying to receive reports, same thing (maybe check OS permissions?).

**Reply by 2xxn** [reply]
Well I gave too few information on what I want to achieve, I wanna build a webbased driver for controlling an RP2040 macropad where I use WebHID to connect to the HID device itself, I'm just wondering how can I register a report descriptor in the firmware that will get handled but I'm not sure if your arduino pico core provides such capabilities

**Reply by earlephilhower** [reply]
Use the `Adafruit TinyUSB` library.  You can make your own descriptors and reports at a low level.  It's not compatible with the `Keyboard` library but you can write your own keyboard HID device relatively simply with `Adafruit TinyUSB`.

**Reply by 2xxn** [reply]
Adafruit TinyUSB seems to work flawlessly, thanks for the recommendation
